specs/GetSecurityDescriptorForAppContainerNames/GetSecurityDescriptorForAppContainerNames.md

```diff
@@ -0,0 +1,87 @@
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_security_descriptor sd;
+    AppContainerAccess access[1] =
+      {{L"Contoso.Test.App_12345678",
+        EVENT_MODIFY_STATE | SYNCHRONIZE}};
+
+    THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
+        1, access, EVENT_MODIFY_STATE | SYNCHRONIZE, &sd);
+
+    SECURITY_ATTRIBUTES sa;
+    sa.nLength = sizeof(sa);
+    sa.lpSecurityDescriptor = sd.get();
+    sa.bInheritHandle = FALSE;
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, &sa);
+    // aka, ::CreateEvent(&sa, FALSE, FALSE, name);
+    return result;
+}
+```
+
+# API Details
+
+```c
+struct AppContainerAccess
+{
+    PCWSTR appContainerName;
+    uint32_t accessMask;
+};
+
+STDAPI GetSecurityDescriptorForAppContainerNames(
+    uint32_t countOfAppContainerNames,
+    _In_reads_(countOfAppContainerNames)
+        const AppContainerAccess* appAccess,
+    uint32_t userAccessMask,
+        _Outptr_ PSECURITY_DESCRIPTOR* securityDescriptor
+)
+```
+
+If the function succeds, the returned `SECURITY_DESCRIPTOR` must be freed by calling [LocalFree](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-localfree).
+
+# API Notes
+
+Some alternative designs were considered.
+- Single method, takes a list of permissions corresponding to the list of PFNs. (Current choice)
+  - Pros
+    - Handles both complicated case and simple case in a single call.
+    - API is straightforward, no extra/optional/multimode parameters.
+    - Only a single method to search/remember/code.
+  - Cons
+    - Simple case is not "optimally" simple - requires filling array with a single value.
+- Single method, takes both a single permission as well as an optional array of permissions. The single permission is ignored if the array is non-null.
+  - Pros
+    - Handles both complicated case and simple case in a single call.
+    - Only a single method to search/remember/code.
+    - Provides a way to assign a single permission to all PFNs without filling an array.
+  - Cons
+    - Slightly more complicated API. Two parameters are conditionally used for the same purpose.
+- Two methods, one that takes a list of permissions, and one that takes a single permission.
+  - Pros
+    - Each API is straightforward, no extra/optional/multimode parameters.
+    - Dedicated APIs for assigning a single permission vs different permissions.
```

Comment:
I'm okay with adding a second method later if necessary. It would be called `GetSecurityDescriptorForAppContainerName` (singular).

Reply:
DISCUSSION: Agreed, we'll ask the customers